pkg/v1/endpoints.go

```diff
@@ -0,0 +1,346 @@
+				klog.Infof("++ volume[%v]: %v", i, vol)
+				mappings = append(mappings, vol)
+			}
+		}
+	}
+
+	return mappings, status, err
+}
+
+// ShowSnapshots : Show one snaphot, or all snapshots, or all snapshots for a volume
+func (client *Client) ShowSnapshots(snapshotId string, sourceVolumeId string) ([]common.SnapshotObject, *common.ResponseStatus, error) {
+
+	request := ""
+	returnSnapshots := []common.SnapshotObject{}
+
+	// Create the correctly formatted request command
+	if sourceVolumeId != "" {
+		request = fmt.Sprintf("/show/snapshots/volume/%q", sourceVolumeId)
+	} else if snapshotId != "" {
+		request = fmt.Sprintf("/show/snapshots/pattern/%q", snapshotId)
+	} else {
+		request = "/show/snapshots"
+	}
+
+	data, status, err := client.FormattedRequest(request)
+
+	// Fill in Snapshot properties for all data objects returned
+	if err == nil && status.ResponseTypeNumeric == 0 {
+		for _, object := range data.Objects {
+			if object.Name == "snapshots" {
+				snapshot := common.SnapshotObject{}
+				snapshot.ObjectName = object.Name
+				snapshot.CreationTime, _ = common.CreationTimeFromString(object.PropertiesMap["creation-date-time-numeric"].Data)
+				snapshot.CreationDateTime = object.PropertiesMap["creation-date-time"].Data
+				snapshot.CreationDateTimeNumeric, _ = strconv.ParseInt(object.PropertiesMap["creation-date-time-numeric"].Data, 10, 64)
+				snapshot.MasterVolumeName = object.PropertiesMap["master-volume-name"].Data
+				snapshot.Name = object.PropertiesMap["name"].Data
+				snapshot.StoragePoolName = object.PropertiesMap["storage-pool-name"].Data
+				snapshot.TotalSize = object.PropertiesMap["total-size"].Data
+				snapshot.TotalSizeNumeric, _ = strconv.ParseInt(object.PropertiesMap["total-size-numeric"].Data, 10, 64)
+				snapshot.VolumeParent = object.PropertiesMap["volume-parent"].Data
+
+				returnSnapshots = append(returnSnapshots, snapshot)
+			}
+		}
+	}
+
+	return returnSnapshots, status, err
+}
+
+// CreateSnapshot : create a snapshot in a snap pool and the snap pool if it doesn't exsits
+func (client *Client) CreateSnapshot(name string, snapshotName string) (*common.ResponseStatus, error) {
+	_, status, err := client.FormattedRequest("/create/snapshots/volumes/%q/%q", name, snapshotName)
+	return status, err
+}
+
+// DeleteSnapshot : delete a snapshot
+func (client *Client) DeleteSnapshot(names ...string) (*common.ResponseStatus, error) {
+	_, status, err := client.FormattedRequest("/delete/snapshot/%q", strings.Join(names, ","))
+	return status, err
+}
+
+// CopyVolume : create an new volume by copying another one or a snapshot
+func (client *Client) CopyVolume(sourceName string, destinationName string, pool string) (*common.ResponseStatus, error) {
+	_, status, err := client.FormattedRequest("/copy/volume/destination-pool/%q/name/%q/%q", pool, destinationName, sourceName)
+	return status, err
+}
+
+func (client *Client) GetVolumeMapsHostNames(name string) ([]string, *common.ResponseStatus, error) {
```

Comment:
I think this may need some more scrutiny due to the way the response can change depending on how host groups are set up.

Reply:
Agreed. This is a v1 feature that I was not focused on changing.